xsd/bcsd.py

```diff
@@ -70,55 +66,94 @@ def make_source_grid(obj):
-    da_obs = make_source_grid(da_obs)
-    da_train = make_source_grid(da_train)
-    da_predict = make_source_grid(da_predict)
+
+def bcsd(da_obs, da_train, da_predict, var='pr'):
+    ''' Apply the Bias Correction and Statistical Downscaling (BCSD) method.
+
+    Parameters
+    ----------
+    da_obs : xr.DataArray
+        Array representing the observed (truth) values.
+    da_train : xr.DataArray
+        Array representing the training data.
+    da_predict : xr.DataArray
+        Array representing the prediction data to be corrected using the BCSD
+        method.
+    var : str
+        Variable name triggering particular treatment of some variables. Valid
+        options include {'pr', 'tmin', 'tmax', 'trange', 'tavg'}.
+
+    Returns
+    -------
+    out_regrid : xr.DataArray
+        Anomalies on the same grid as ``da_obs``.
+    '''
 
     # regrid to common course grid
     bounds = get_bounds(da_obs)
-    course_grid = course_grid = xe.util.grid_2d(*bounds['lon'], 1, *bounds['lat'], 1)
-    regridder = xe.Regridder(da_obs, course_grid, 'bilinear')
-    da_obs_regrid = regridder(da_obs)
+    course_grid = xe.util.grid_2d(*bounds['lon'], 1, *bounds['lat'], 1)
+    da_obs_regrid, da_train_regrid, da_predict_regrid = _regrid_to(
+        course_grid, da_obs, da_train, da_predict)
 
-    # regrid training/predict data to common grid
-    regridder = xe.Regridder(da_train, course_grid, 'bilinear')
-    da_train_regrid = regridder(da_train)
-    da_predict_regrid = regridder(da_predict)
-
-    # Calc means for train/paedict
+    # Calc mean climatology for training data
     da_train_regrid_mean = da_train_regrid.groupby(
         'time.month').mean(dim='time')
 
     if var == 'pr':
         # Bias correction
-        # apply quantile mapping
+        # apply quantile mapping by month
         da_predict_regrid_qm = quantile_mapping_by_group(
             da_predict_regrid, da_train_regrid, da_obs_regrid,
             grouper='time.month')
 
-        da_predict_regrid_anoms = da_predict_regrid_qm.groupby('time.month') / da_train_regrid_mean
+        # calculate the amonalies as a ratio of the training data
+        # again, this is done month-by-month
+        da_predict_regrid_anoms = (da_predict_regrid_qm.groupby('time.month')
+                                   / da_train_regrid_mean)
     else:
-        # don't do this for training period? check with andy
-        da_predict_regrid_run = da_predict_regrid.groupby(
-            'time.month').rolling(time='9Y', center=True).mean('time')
-        # what do we do with these?
-        da_predict_changes = da_predict_regrid - da_predict_regrid_run
+        # Calculate the 9-year running mean for each month
+        # Q: don't do this for training period? check with andy
+        da_predict_regrid_rolling_mean = da_predict_regrid.groupby(
+            'time.month').apply(_running_mean, time=9, center=True,
+                                min_periods=1)
```

Comment:
@andywood - do you recall what was done for the edges of the record when calculating the rolling mean? This line executes a rolling mean for each month, with a window size of 9 years. The window is centered on the year in question (e.g. +/- 4 years). I have set `min_periods` to 1 so we get values for all years, even at the beginning and end of the record. This is the main thing I have questions about but perhaps other details here raise flags in your mind.

Reply:
I think I just progressively shortened the averaging period to include whatever data were in the 'centered' window. So the first day mean was only over the 1st 5 days.  When I get my laptop back (it died last week) I'll check the code, unless someone has a copy.  It's a bit of perl.